### PR TITLE
feat: lotus-provider: WinningPoSt support

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
Closes https://github.com/filecoin-project/lotus/issues/11392
Closes https://github.com/filecoin-project/lotus/issues/11383

## Additional Info
* Fully implemented afaict
* 100% untested

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: lib/harmony/harmonydb/sql/20231110.sql [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+(
+    id        bigserial not null
+        constraint mining_base_block_pk
+            primary key,
+    task_id   bigint    not null
+        constraint mining_base_block_mining_tasks_task_id_fk
+            references mining_tasks
+            on delete cascade,
+    block_cid text      not null
+        constraint mining_base_block_cid_k
```
Comment (magik6k) [inline lib/harmony/harmonydb/sql/20231110.sql:31]: This constraint also need sp_id

File: lib/harmony/harmonydb/sql/20231110.sql [hunk 2]
```diff
@@ -0,0 +1,39 @@
+create table mining_tasks
```
Comment (mtrisic) [inline lib/harmony/harmonydb/sql/20231110.sql:1]: Should we maybe also persist info about which instance executed the task?  🤔 That might be helpful in analyzing and optimising multi-worker environment

Comment (snadrus) [reply lib/harmony/harmonydb/sql/20231110.sql:1]: We have the completed tasks table for harmonytask. We could add the field there and update it for all completed tasks.

File: miner/miner.go [hunk 3]
```diff
@@ -312,19 +325,23 @@ minerLoop:
...
 					build.Clock.Sleep(build.Clock.Until(btime))
 				}
 			default:
+				// Log if the block was mined in the past.
 				log.Warnw("mined block in the past",
 					"block-time", btime, "time", build.Clock.Now(), "difference", build.Clock.Since(btime))
 			}
 
+			// Check for slash filter conditions.
 			if os.Getenv("LOTUS_MINER_NO_SLASHFILTER") != "_yes_i_know_i_can_and_probably_will_lose_all_my_fil_and_power_" && !build.IsNearUpgrade(base.TipSet.Height(), build.UpgradeWatermelonFixHeight) {
```
Comment (snadrus) [inline miner/miner.go:345]: Should this os.GetEnv() be replaced with a config in the database?

**Review by snadrus — APPROVED** [review]
minor nit about config consistency

File: cmd/lotus-provider/migrate.go [hunk 4]
```diff
@@ -161,7 +175,7 @@ environment variable LOTUS_WORKER_WINDOWPOST.
 	}
 
 	if !lo.Contains(titles, "base") {
-		_, err = db.Exec(ctx, "INSERT INTO harmony_config (title, config) VALUES ('base', '')", "base")
+		_, err = db.Exec(ctx, "INSERT INTO harmony_config (title, config) VALUES ('base', '')")
```
Comment (snadrus) [inline cmd/lotus-provider/migrate.go:178]: We probably should have the migrator enable:
lpCfg.Subsystems.EnableWinningPost = true
And add a comment about diabling it in lotus-miner

**Comment by snadrus** [comment]
Orjan mentioned: "null-blocks and a couple of other blockchain edge-cases". Be sure that's covered before merging since feat/sturdypost is planned for 1.25.1.

**Pull request merged**